src/book.py

```diff
@@ -180,39 +217,79 @@ def tokenize(self, on, rem_stopwords=True, stopword_lang='english',
+        non-cont: (Default = False): If True, returns random char/word/sent from starting point
+        randomized: (Default = False): If True, chooses a random point to start making snippets
+        rem_stopwords: Same as in Tokenize
+        ret_as_array: (Default = False): Returns one BookText object by appending all groups of snippets, else returns
+                                         an array of BookText objects
+        random_seed: When positive, generates the same snippets everytime
+        
         """
-        # TODO add ability for random seed
+        if (random_seed is not None):
+            random.seed(random_seed)
+            
         if 'char' in on.lower():
-            assert length < len(self._text)
+            tokens = self._text
         elif 'word' in on.lower():
-            tokens = word_tokenize(self._text)
-            assert length < len(tokens)
+            tokens = self.tokenize('word', rem_stopwords)
         elif 'sent' in on.lower():
-            tokens = sent_tokenize(self._text)
+            tokens = self.tokenize('sent', rem_stopwords)
         else:
             raise KeyError(
-                "Arugument 'on' must refer to character, word, or sentence")
-
-        if non_cont:
-            # TODO add functionality for non-continuous random sampling
-            pass
+                "Argument 'on' must refer to character, word, or sentence")
+            
+        assert groups*length <= len(tokens)
+        return_array = []
+
+        if randomized:
+            start = random.randint(0, len(tokens) - groups*length)
+            tokens = tokens[start:]
+            start = 0
         else:
-            if 'char' in on.lower():
-                start = random.randint(0, len(self._text) - length)
-                snippet = self._text[start:start + length]
-            elif 'word' in on.lower():
-                start = random.randint(0, len(tokens) - length)
-                snippet = (" ").join(tokens[start:start + length])
-            elif 'sent' in on.lower():
-                start = random.randint(0, len(tokens) - length)
-                snippet = (" ").join(tokens[start:start + length])
+            start = 0
+        
+        if non_cont == False:
+            for gr in range(groups):
+                snippet = BookText(rawtext=''.join((" ").join(tokens[length*gr:length*(gr+1)])), 
+                               author = self.author, title = self.title, meta = self.meta)
+                return_array.append(snippet)
+        else:
+            for gr in range(groups): 
+                snippet = BookText(rawtext=''.join((" ").join(tokens[start:start+length])), 
+                               author = self.author, title = self.title, meta = self.meta)
+                snippet._text.rstrip()    
+                return_array.append(snippet)
+                
+                for index in sorted(range(start, start+length), reverse=True):
+                    if (len(tokens)-length > 0):
```

Comment:
I'm unclear what this line is doing. Is tokens being updated at some point in this loop?

Reply:
Yes, to avoid repetition between snippets in a group of snippets, I drop the char/word/sent which already appeared.

Reply:
But won't this give you either less groups than you asked for or a snippet which was not of the length requested?

Reply:
But won't `len(tokens)` always be greater than 0? So this step is essentially always going off? Haven't you already ensured that you won't get the exact same snippet twice by having your index set by `sorted(range(start, start+length)`?

Reply:
it's len(tokens) - length. That changes at every iteration. This bit runs only when you have (say) 9 sentences, and you want 3 groups of 3 sentences each. In that case, this bit is satisfied at the last step.

Reply:
But to avoid repetition of snippets you only ever need to delete the first element in `tokens`. It should be ok if we ask for a 2 snippets and the first one is all of chapter 1 and the second one is all of chapter 1 minus the first word plus the first word of chapter 2.

Reply:
After in-person conversations, we were both looking for different implementations of this function. This disagreement is less important than getting this code out so I will add this as an "issue" to be resolved later.